<?xml version="1.0" ?>
<robot xmlns:xacro="http://ros.org/wiki/xacro" name="tflow">

    <xacro:property name="dynamixel_mx28_x" value="0.0356"/>
    <xacro:property name="dynamixel_mx28_y" value="0.0355"/>
    <xacro:property name="dynamixel_mx28_z" value="0.0506"/>
    <xacro:property name="dynamixel_mx28_weight" value="0.072"/>

    <xacro:property name="dynamixel_mx64_x" value="0.0402"/>
    <xacro:property name="dynamixel_mx64_y" value="0.041"/>
    <xacro:property name="dynamixel_mx64_z" value="0.0611"/>
    <xacro:property name="dynamixel_mx64_weight" value="0.126"/>

    <xacro:property name="dynamixel_mx106_x" value="0.0402"/>
    <xacro:property name="dynamixel_mx106_y" value="0.046"/>
    <xacro:property name="dynamixel_mx106_z" value="0.0651"/>
    <xacro:property name="dynamixel_mx106_weight" value="0.153"/>

    <xacro:property name="kepalakepala_y" value="0.046"/>
    <xacro:property name="kepalakepala_z" value="0.0651"/>
    <xacro:property name="kepalakepala_weight" value="0.153"/>

    <xacro:macro name="kepalakepala" params="roll pitch yaw x y z">
        <visual>
            <geometry>
                <box size="${kepalakepala_x} ${kepalakepala_y} ${kepalakepala_z}"/>
            </geometry>
            <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
            <material name="silver">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
          <geometry>
            <box size="${kepalakepala_x} ${kepalakepala_y} ${kepalakepala_z}"/>
          </geometry>
          <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
        </collision>
        <inertial>
            <mass value="${kepalakepala_weight}"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
    </xacro:macro>

    <xacro:macro name="dynamixelmx28" params="roll pitch yaw x y z">
        <visual>
            <geometry>
                <box size="${dynamixel_mx28_x} ${dynamixel_mx28_y} ${dynamixel_mx28_z}"/>
            </geometry>
            <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
            <material name="silver">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
          <geometry>
            <box size="${dynamixel_mx28_x} ${dynamixel_mx28_y} ${dynamixel_mx28_z}"/>
          </geometry>
          <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
        </collision>
        <inertial>
            <mass value="${dynamixel_mx28_weight}"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
    </xacro:macro>

    <xacro:macro name="dynamixelmx64" params="roll pitch yaw x y z">
        <visual>
            <geometry>
                <box size="${dynamixel_mx64_x} ${dynamixel_mx64_y} ${dynamixel_mx64_z}"/>
            </geometry>
            <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
            <material name="silver">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
          <geometry>
            <box size="${dynamixel_mx64_x} ${dynamixel_mx64_y} ${dynamixel_mx64_z}"/>
          </geometry>
          <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
        </collision>
        <inertial>
            <mass value="${dynamixel_mx64_weight}"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
    </xacro:macro>

    <xacro:macro name="dynamixelmx106" params="roll pitch yaw x y z">
        <visual>
            <geometry>
                <box size="${dynamixel_mx106_x} ${dynamixel_mx106_y} ${dynamixel_mx106_z}"/>
            </geometry>
            <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
            <material name="silver">
                <color rgba="0.3 0.3 0.3 1"/>
            </material>
        </visual>
        <collision>
          <geometry>
            <box size="${dynamixel_mx106_x} ${dynamixel_mx106_y} ${dynamixel_mx106_z}"/>
          </geometry>
          <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
        </collision>
        <inertial>
            <mass value="${dynamixel_mx106_weight}"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
    </xacro:macro>

    <xacro:macro name="link_cylinder" params="le r roll pitch yaw x y z">
        <visual>
            <geometry>
                <cylinder length="${le}" radius="${r}"/>
            </geometry>
            <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
            <material name="white">
                <color rgba="0.8 0.8 0.8 1"/>
            </material>
        </visual>
        <collision>
          <geometry>
            <cylinder length="${le}" radius="${r}"/>
          </geometry>
          <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
        </collision>
        <inertial>
            <mass value="${le*r}"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
    </xacro:macro>

    <xacro:macro name="link_box" params="bx by bz roll pitch yaw x y z">
        <visual>
            <geometry>
                <box size="${bx} ${by} ${bz}"/>
            </geometry>
            <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
            <material name="black">
                <color rgba="0 0 0 1"/>
            </material>
        </visual>
        <collision>
          <geometry>
            <box size="${bx} ${by} ${bz}"/>
          </geometry>
          <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
        </collision>
        <inertial>
            <mass value="${bx*by*bz}"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
    </xacro:macro>
    
     <xacro:macro name="link_box2" params="bx by bz roll pitch yaw x y z">
        <visual>
            <geometry>
                <box size="${bx} ${by} ${bz}"/>
            </geometry>
            <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
            <material name="white">
                <color rgba="0.8 0.8 0.8 1"/>
            </material>
        </visual>
        <collision>
          <geometry>
            <box size="${bx} ${by} ${bz}"/>
          </geometry>
          <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
        </collision>
        <inertial>
            <mass value="${bx*by*bz}"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
    </xacro:macro>

    <xacro:macro name="telapak_tangan" params="le r roll pitch yaw x y z">
        <visual>
            <geometry>
                <cylinder length="${le}" radius="${r}"/>
            </geometry>
            <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
            <material name="black">
                <color rgba="0 0 0 1"/>
            </material>
        </visual>
        <collision>
          <geometry>
            <cylinder length="${le}" radius="${r}"/>
          </geometry>
          <origin rpy="${roll} ${pitch} ${yaw}" xyz="${x} ${y} ${z}"/>
        </collision>
        <inertial>
            <mass value="${le*r}"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
    </xacro:macro>

    <link name="base_link">
        <visual>
            <geometry>
                <box size="0.18 0.12 0.08"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="silver">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <box size="0.18 0.12 0.08"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="silver">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </collision>
        <inertial>
            <mass value="0.01"/>
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
        </inertial>
    </link>

    <!-- PERUT -->
    <link name="servo_perut_roll_1">
        <xacro:dynamixelmx106 roll="0.0" pitch="0.0" yaw = "1.57" x="0" y="0" z="0"/>
    </link>

    <link name="servo_perut_roll_2">
        <xacro:dynamixelmx106 roll="0.0" pitch="0.0" yaw = "-1.57" x="0" y="0" z="0"/>
    </link>

    <link name="servo_perut_pitch">
        <xacro:dynamixelmx106 roll="0.0" pitch="0.0" yaw = "0" x="0" y="0" z="0"/>
    </link>

    <link name="link_perut_yaw">
        <xacro:link_cylinder le="0.04" r="0.02" roll="0.0" pitch="0.0" yaw ="0" x="0" y="0" z="0.03"/>
    </link>

    <link name="servo_perut_yaw">
        <xacro:dynamixelmx106 roll="-1.57" pitch="0.0" yaw = "0" x="0" y="0" z="0.02"/>
    </link>
    <!-- PERUT_END -->

    <!-- Kepala -->
    	
    <link name="servo_kepala_yaw">
        <xacro:dynamixelmx28 roll="1.57" pitch="0.0" yaw = "0" x="0" y="0" z="0.0"/>
    </link>
    
    <link name="link_leher">
    	<xacro:link_cylinder le="0.031" r="0.015" roll="0.0" pitch="0.0" yaw ="0" x="0" y="0" z="0.0	"/>
   	</link>
    
    <link name="servo_kepala_pitch">
        <xacro:dynamixelmx28 roll="0.00" pitch="0.0" yaw = "0" x="0" y="-0.04" z="0.0"/>
    </link>
    
    <link name="servo_kepala_roll">
        <xacro:dynamixelmx28 roll="0.00" pitch="0.0" yaw = "0" x="0" y="0.00" z="0.0"/>
    </link>
    
    <link name="kepala">
    	<xacro:link_box2 bx="0.125" by="0.085" bz="0.042" roll="0.0" pitch="0.0" yaw ="0" x="0" y="0" z="0.07	"/>
   	</link>
    
    
    <!-- Kepala_END -->

    <!-- Tangan_Kanan -->
    <link name="servo_bahu_kanan_roll">
        <xacro:dynamixelmx106 roll="-1.57" pitch="0" yaw = "0" x="0" y="0" z="0"/>
    </link>

    <link name="servo_bahu_kanan_pitch">
        <xacro:dynamixelmx106 roll="0" pitch="1.57" yaw = "0" x="0" y="0" z="0"/>
    </link>

    <link name="link_siku_kanan">
        <xacro:link_cylinder le="0.195" r="0.01" roll="0.0" pitch="1.57" yaw ="0" x="${0.195/2}" y="0" z="0.0"/>
    </link>

    <link name="servo_siku_kanan_pitch">
        <xacro:dynamixelmx106 roll="0" pitch="1.57" yaw = "1.57" x="0" y="0" z="0"/>
    </link>

    <link name="link_lengan_kanan">
        <xacro:link_cylinder le="0.15" r="0.01" roll="0.0" pitch="1.57" yaw ="0" x="${-(0.16/2)}" y="0" z="0.0"/>
    </link>

    <link name="link_telapak_kanan">
        <xacro:link_box bx="0.015" by="0.07" bz="0.04" roll="0.0" pitch="0" yaw ="0" x="0" y="0" z="0.0"/>
    </link>

    <link name="telapak_tangan_kanan">
        <xacro:telapak_tangan le="0.01" r="${0.125/2}" roll="0.0" pitch="0" yaw ="0" x="0" y="0" z="0.0"/>
    </link>
    <!-- Tangan_Kanan_END -->

    <!-- Tangan_Kiri -->
    <link name="servo_bahu_kiri_roll">
        <xacro:dynamixelmx106 roll="1.57" pitch="0" yaw = "0" x="0" y="0" z="0"/>
    </link>

    <link name="servo_bahu_kiri_pitch">
        <xacro:dynamixelmx106 roll="0" pitch="-1.57" yaw = "0" x="0" y="0" z="0"/>
    </link>

    <link name="link_siku_kiri">
        <xacro:link_cylinder le="0.195" r="0.01" roll="0.0" pitch="-1.57" yaw ="0" x="${-(0.195/2)}" y="0" z="0.0"/>
    </link>

    <link name="servo_siku_kiri_pitch">
        <xacro:dynamixelmx106 roll="0" pitch="1.57" yaw = "1.57" x="0" y="0" z="0"/>
    </link>

    <link name="link_lengan_kiri">
        <xacro:link_cylinder le="0.15" r="0.01" roll="0.0" pitch="-1.57" yaw ="0" x="${(0.16/2)}" y="0" z="0.0"/>
    </link>

    <link name="link_telapak_kiri">
        <xacro:link_box bx="0.015" by="0.07" bz="0.04" roll="0.0" pitch="0" yaw ="0" x="0" y="0" z="0.0"/>
    </link>

    <link name="telapak_tangan_kiri">
        <xacro:telapak_tangan le="0.01" r="${0.125/2}" roll="0.0" pitch="0" yaw ="0" x="0" y="0" z="0.0"/>
    </link>
    <!-- Tangan_Kiri_END -->
    
     <!-- Kepala -->
    <joint name="perut_to_kepala" type="fixed">
        <parent link="servo_perut_yaw"/>
        <child link="servo_kepala_yaw"/> <!--servo_kepala_roll-->
        <origin rpy="0 0 0" xyz="0 0.0 0.12"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 0 1" />
    </joint>
    
    <joint name="kepala_yaw" type="revolute">
        <parent link="servo_kepala_yaw"/>
        <child link="link_leher"/> <!--servo_kepala_roll-->
        <origin rpy="0 0 0" xyz="0 0.0 0.02"/>
        <limit effort="300" velocity="0.1" lower="-3.14" upper="3.14"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 0 1" />
    </joint>
    
    <joint name="kepala_pitch" type="revolute">
        <parent link="link_leher"/>
        <child link="servo_kepala_pitch"/> <!--servo_kepala_roll-->
        <origin rpy="0 0 0" xyz="0 0.0 0.045"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 -1 0" />
    </joint>
    
    <joint name="joint_kepala_yaw" type="fixed">
        <parent link="servo_kepala_pitch"/>
        <child link="servo_kepala_roll"/> <!--servo_kepala_roll-->
        <origin rpy="0 0 0" xyz="0 0.005 0.0"/>
        <limit effort="300" velocity="0.1" lower="-3.14" upper="3.14"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 0 1" />
    </joint>
    
    <joint name="kepala_roll" type="revolute">
        <parent link="servo_kepala_roll"/>
        <child link="kepala"/> <!--servo_kepala_roll-->
        <origin rpy="0 0 0" xyz="0 0.0 0.0"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="-1 0 0" />
    </joint>
        
    <!-- Kepala_END -->
    

    <!-- PERUT -->
    <joint name="joint_perut_roll" type="fixed">
        <parent link="base_link"/>
        <child link="servo_perut_roll_1"/>
        <origin rpy="0 0 0" xyz="0.055 0.0 0.1"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="1 0 0" />
    </joint>

    <joint name="joint_perut_roll_2" type="fixed">
        <parent link="base_link"/>
        <child link="servo_perut_roll_2"/>
        <origin rpy="0 0 0" xyz="-0.055 0.0 0.1"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="1 0 0" />
    </joint>

    <joint name="perut_roll" type="revolute">
        <parent link="servo_perut_roll_1"/>
        <child link="servo_perut_pitch"/>
        <origin rpy="0 0 0" xyz="-0.055 0.0 0.05"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="1 0 0" />
    </joint>

    <joint name="perut_pitch" type="revolute">
        <parent link="servo_perut_pitch"/>
        <child link="link_perut_yaw"/>
        <origin rpy="0 0 0" xyz="0 0.0 0.0"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 1 0" />
    </joint>

    <joint name="perut_yaw" type="revolute">
        <parent link="link_perut_yaw"/>
        <child link="servo_perut_yaw"/>
        <origin rpy="0 0 0" xyz="0.0 0.0 0.05"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 0 1" />
    </joint>

    <!-- PERUT_END -->

    <!-- Tangan -->

    <joint name="joint_bahu_kiri_yaw" type="fixed">
        <parent link="servo_kepala_yaw"/>
        <child link="servo_bahu_kiri_roll"/>
        <origin rpy="0 0 0" xyz="-0.08 0.0 0.0"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 0 1" />
        </joint>

    <joint name="bahu_kiri_roll" type="revolute">
        <parent link="servo_bahu_kiri_roll"/>
        <child link="servo_bahu_kiri_pitch"/>
        <origin rpy="0 0 0" xyz="-0.065 0.0 0.0015"/>
        <limit effort="300" velocity="0.1" lower="-3.14" upper="3.14"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="-1 0 0" />
    </joint>

    <joint name="joint_bahu_kanan_roll" type="fixed">
        <parent link="servo_kepala_yaw"/>
        <child link="servo_bahu_kanan_roll"/>
        <origin rpy="0 0 0" xyz="0.08 0.0 0.0"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 0 1" />
    </joint>

    <joint name="bahu_kanan_roll" type="revolute">
        <parent link="servo_bahu_kanan_roll"/>
        <child link="servo_bahu_kanan_pitch"/>
        <origin rpy="0 0 0" xyz="0.065 0.0 0.0015"/>
        <limit effort="300" velocity="0.1" lower="-3.14" upper="3.14"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="1 0 0" />
    </joint>

    <joint name="bahu_kiri_yaw" type="revolute">
        <parent link="servo_bahu_kiri_pitch"/>
        <child link="link_siku_kiri"/>
        <origin rpy="0 -1.57 0" xyz="-0.03 0.0 0.00"/>
        <limit effort="300" velocity="0.1" lower="-2.00" upper="2.00"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 1 0" />
    </joint>

    <joint name="bahu_kanan_yaw" type="revolute">
        <parent link="servo_bahu_kanan_pitch"/>
        <child link="link_siku_kanan"/>
        <origin rpy="0 1.57 0" xyz="0.03 0.0 0.00"/>
        <limit effort="300" velocity="0.1" lower="-2.00" upper="2.00"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 1 0" />
    </joint>

    <joint name="siku_kiri_yaw" type="revolute">
        <parent link="link_siku_kiri"/>
        <child link="servo_siku_kiri_pitch"/>
        <origin rpy="0 0 0" xyz="-0.195 0.0 0.00"/>
        <limit effort="300" velocity="0.1" lower="-2.00" upper="2.00"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 0 -1" />
    </joint>

    <joint name="siku_kanan_yaw" type="revolute">
        <parent link="link_siku_kanan"/>
        <child link="servo_siku_kanan_pitch"/>
        <origin rpy="0 0 0" xyz="0.195 0.0 0.00"/>
        <limit effort="300" velocity="0.1" lower="-2.00" upper="2.00"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 0 -1" />
    </joint>

    <joint name="joint_lengan_kiri_yaw" type="fixed">
        <parent link="servo_siku_kiri_pitch"/>
        <child link="link_lengan_kiri"/>
        <origin rpy="0 0 0" xyz="-0.16 0.0 0.00"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 0 1" />
    </joint>

    <joint name="joint_tangan_kiri" type="fixed">
        <parent link="link_lengan_kiri"/>
        <child link="link_telapak_kiri"/>
        <origin rpy="0 0 0" xyz="${-(0.015/2)} 0.0 0.0"/>
        <limit effort="300" velocity="0.1" lower="-3.14" upper="3.14"/>
        <dynamics damping="50" friction="1"/>
    </joint>

    <joint name="tangan_kiri_EOF" type="fixed">
        <parent link="link_telapak_kiri"/>
        <child link="telapak_tangan_kiri"/>
        <origin rpy="0 0 0" xyz="${-(0.125/2)} 0.0 0.02"/>
        <limit effort="300" velocity="0.1" lower="-3.14" upper="3.14"/>
        <dynamics damping="50" friction="1"/>
    </joint>

    <joint name="joint_lengan_kanan_yaw" type="fixed">
        <parent link="servo_siku_kanan_pitch"/>
        <child link="link_lengan_kanan"/>
        <origin rpy="0 0 0" xyz="0.16 0.0 0.00"/>
        <limit effort="300" velocity="0.1" lower="-1.57" upper="1.57"/>
        <dynamics damping="50" friction="1"/>
        <axis xyz="0 0 1" />
    </joint>

    <joint name="joint_tangan_kanan" type="fixed">
        <parent link="link_lengan_kanan"/>
        <child link="link_telapak_kanan"/>
        <origin rpy="0 0 0" xyz="${0.015/2} 0.0 0.0"/>
        <limit effort="300" velocity="0.1" lower="-3.14" upper="3.14"/>
        <dynamics damping="50" friction="1"/>
    </joint>

    <joint name="tangan_kanan_EOF" type="fixed">
        <parent link="link_telapak_kanan"/>
        <child link="telapak_tangan_kanan"/>
        <origin rpy="0 0 0" xyz="${0.125/2} 0.0 0.02"/>
        <limit effort="300" velocity="0.1" lower="-3.14" upper="3.14"/>
        <dynamics damping="50" friction="1"/>
    </joint>


</robot>
